### ISSUE TITLE
tail -f for logs

### ISSUE THREAD

**Issue by jkustan** [jtmoon79/super-speedy-syslog-searcher#142]
## Summary

This is an impressive project! We've started using it on several projects. 

It would be super useful to be able to tail these consolidated logs "-f", i.e. see changes as they come in, including any greps etc that we have. Is this already possible / could it be added?

### Current behavior

s4 searches current logs and returns.

### Suggested behavior

s4 -t or --tail: Doesn't return, just continues to print logs as they are written to the disk

**Comment by jtmoon79** [comment]
Glad it's useful! 😁

I'll have to think about this as it's not a trivial change. It breaks some foundational assumptions I started with. But if you're running into a need for it then I'll seriously consider it.

**Comment by jkustan** [comment]
Thanks! The other solution I thought of was to write a shell script that
would run every second, passing -a=[time of last report]. It won't be
exactly a constant flow, but it will be pretty good.

The only issue I ran into in doing a PoC was that the -a parameter doesn't
seem to parse the format that includes a fractional second.

This time parses: 20230626T075830
but this one does not: 20230626T075830.996

If I want to get this right, I'll need to parse the fractional second so it
picks up exactly where it left off.

Just some thoughts -- thanks for the response!

**Comment by jtmoon79** [comment]
***tl;dr*** please describe the scenario that motivated this request.

Now that I've thought about this, I recall I decided to never support such a scenario, real-time following log files like `tail -f`, because this specific scenario doesn't make any sense. If log files are being followed in real-time then just print log messages as they arrive using `tail -f`. There may be an unpredictable delay between log message generation and delivery to some logging system endpoint, but that delay should be small enough to ignore. So I don't understand how `s4` could help here.

To consider this another way, if log files are receiving log messages at nearly the same real time but with wildly different datetimestamps, so widly different that so you need `s4` to resort them by datetimestamp, then something is wrong with that logging system.

So I'm unable to imagine your scenario that motivated this request. Could you please describe it in detail?

**Comment by jtmoon79** [comment]
> The only issue I ran into in doing a PoC was that the -a parameter doesn't
seem to parse the format that includes a fractional second.

See #145

**Comment by jkustan** [comment]
Hi James,

I'm trying to create the illusion of 30 separate servers writing to
individual log files on a shared AWS EFS drive, each in their own
directory, coming through as if it were a single virtual server with
monolithic log file that I can just "tail". So what we did in the past with
a single server was tail -f | grep 'something' -- then do something on the
website and see only the logs related to the specific thing you are doing.

The issue with tail -f is that it doesn't merge the logs. So:
 tail -f **/*.log

looks like this:
==> log/vendor_logger.log <==
I, [2023-07-03T10:38:01.452020 #77324]  INFO -- : hellodebug

==> log/autodetect_logger.log <==
I, [2023-07-03T10:38:08.061625 #77324]  INFO -- : hellodebug

==> log/development.log <==
hello

where I want it to look like this:
I, [2023-07-03T10:38:01.452020 #77324]  INFO -- : hellodebug
I, [2023-07-03T10:38:08.061625 #77324]  INFO -- : hellodebug
hello

Your (awesome) tool does this on-demand but not in realtime with tail.

I *might* be able to work around the "no fractional time" limitation by
looking for a time where "time > [previous time]" and using grep to filter
from a specific time forward. I'll let you know how things go and if I find
a solution that works, happy to share it with other users.

Thanks,

Jack

**Comment by jtmoon79** [comment]
> The issue with tail -f is that it doesn't merge the logs.

Would [`multitail`](https://www.vanheusden.com/multitail/) with `--mergeall` do what you need?

```text
$ multitail --mergeall -n 0 -ts --follow-all **/*.log
```

**Comment by jtmoon79** [comment]
> > The issue with tail -f is that it doesn't merge the logs.
> 
> Would [`multitail`](https://www.vanheusden.com/multitail/) with `--mergeall` do what you need?
> 
> ```
> $ multitail --mergeall -n 0 -ts --follow-all **/*.log
> ```

@jkustan did that recommendation ☝️  work for you?

**Comment by jkustan** [comment]
I ended up writing a neat little rake task (since I'm better at ruby than shell scripting, works great with one caveat.

    task :tail, [:grep] => :environment do |t, args|
      ARGV.each { |a| task a.to_sym do ; end }

      first = true
      grep = ARGV[1]
      time = nil
    
      while true do
        command = "s4 -nw -c never /app/log/current/ -a="
        command += first ? "-5s": time
        command += " | grep #{grep}" if grep.present?

        tail = `#{command}`
        last_line = tail.split("\n").last

        # each line looks like this: production.log:I, [2023-07-17T04:14:02.139918 #62]  INFO -- : Completed 200 OK in 105ms (Views: 0.4ms | ActiveRecord: 14.8ms | Allocations: 22998)
        time = last_line.split(",")[1].strip.match(/\[(.*)\]/)[1].split(" ").first[0..-4]
        # this will pull out the timestamp, in this case "2023-07-17T04:14:02.139918" Note the 6 decimal places for time
        puts tail
        first = false
      end
    end

Caveat is that the fractional time change seems to only work for 3 decimal places, and since we get sometimes hundreds per second, the log is not continuous. Could you adjust the last fix to allow for 6 decimal places? Then this will work great.

Other caveat is that since it's a rake task it is slow to load, maybe at some point I (or someone else) will convert this to a shell script and it will run without the rails load tax.

Thanks,

Jack

**Comment by jtmoon79** [comment]
I'm still not sure why `multitail` does not fit your needs. But hey, if `s4` is helping out then great! 😃 

Version 0.6.68 has been released which will parse 6-digit fractional seconds in the `-a` and `-b` arguments.

**Comment by jkustan** [comment]
Thank you!! Just what we needed, and I *can* use unix find command to limit
to specific files. All good now --
